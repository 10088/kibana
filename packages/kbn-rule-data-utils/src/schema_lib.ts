/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export interface Validator<T> {
  (value: unknown): value is T;
  explanation(value: unknown): Explanation;
}
type Explanation =
  | [valid: true, explanation: undefined]
  | [valid: false, explanation: (newlinePrefix?: string) => string];
type TypeOf<V extends Validator<unknown>> = V extends Validator<infer T> ? T : never;

/**
 * Validate that `value` matches at least one of `validators`.
 * Use this to create a predicate for a union type.
 * e.g.
 * ```
 * import * as schema from './schema';
 * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
 *   schema.literal('asc' as const),
 *   schema.literal('desc' as const),
 * ]);
 * ```
 */
export function oneOf<V extends Array<Validator<unknown>>>(
  validators: V
): Validator<V extends Array<Validator<infer ElementType>> ? ElementType : never> {
  validator.explanation = function (value: unknown): Explanation {
    const subExplanations: Array<(newlinePrefix?: string) => string> = [];
    for (const subValidator of validators) {
      const result = subValidator.explanation(value);
      if (result[0]) {
        return [true, undefined];
      } else {
        subExplanations.push(result[1]);
      }
    }
    return [
      false,
      (newlinePrefix: string = '') => `${JSON.stringify(
        value,
        null,
        newlinePrefix
      )} was expected to match one of several conditions, but it did not. Here are the explanations:
      ${subExplanations.map((explanation, index) => {
        return `${newlinePrefix + '\t'}* ${explanation(newlinePrefix + '\t\t')}${
          index < subExplanations.length - 1 ? ' OR' : ''
        }`;
      })}`,
    ];
  };
  return validator;
  function validator(
    value: unknown
  ): value is V extends Array<Validator<infer ElementType>> ? ElementType : never {
    for (const subValidator of validators) {
      if (subValidator(value)) {
        return true;
      }
    }
    return false;
  }
}

/**
 * Validate that `value` is an array and that each of its elements matches `elementValidator`.
 * Use this to create a predicate for an array type.
 * ```
 * import * as schema from './schema';
 * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
 *   schema.literal('asc' as const),
 *   schema.literal('desc' as const),
 * ]);
 * ```
 */
export function array<V extends Validator<unknown>>(
  elementValidator: V
): Validator<Array<V extends Validator<infer ElementType> ? ElementType : never>> {
  validator.explanation = function (value: unknown): Explanation {
    // cast Array.isArray to safer type
    if ((Array.isArray as (value: unknown) => value is unknown[])(value)) {
      const explanations: Array<
        [index: number, explanation: (newlinePrefix?: string) => string]
      > = [];
      for (const [index, element] of value.entries()) {
        const result = elementValidator.explanation(element);
        if (!result[0]) {
          explanations.push([index, result[1]]);
        }
      }
      if (explanations.length) {
        return [
          false,
          (newlinePrefix: string = '') =>
            `Values in array were invalid. The explanations: ${explanations.map(
              ([index, explanation]) => {
                const prefix = ``;
                return `
${newlinePrefix}* Value at index ${index}: ${explanation(newlinePrefix + '\t')}`;
              }
            )}`,
        ];
      } else {
        return [true, undefined];
      }
    }
    return [
      false,
      (newlinePrefix: string = '') =>
        `Expected an array, received ${JSON.stringify(value, null, newlinePrefix)}`,
    ];
  };
  return validator;
  function validator(
    value: unknown
  ): value is Array<V extends Validator<infer ElementType> ? ElementType : never> {
    // This validator doesn't delegate to its explanation function because it short circuits instead.
    // cast Array.isArray to safer type
    if ((Array.isArray as (value: unknown) => value is unknown[])(value)) {
      for (const element of value) {
        const result = elementValidator(element);
        if (!result) {
          return false;
        }
      }
      return true;
    }
    return false;
  }
}

/**
 * The keys of `T` where `undefined` is assignable to the corresponding value.
 * Used to figure out which keys could be made optional.
 */
type KeysWithOptionalValues<T extends { [key: string]: unknown }> = {
  [K in keyof T]: undefined extends T[K] ? K : never;
}[keyof T];

/**
 * `T` with required keys changed to optional if the corresponding value could be `undefined`.
 * Converts a type like `{  key: number | undefined; requiredKey: string }` to a type like `{ key?: number | undefined; requiredKey: string }`
 * This allows us to write object literals that omit a key if the value can accept `undefined`.
 */
type OptionalKeyWhenValueAcceptsUndefined<T extends { [key: string]: unknown }> = {
  [K in Exclude<keyof T, KeysWithOptionalValues<T>>]: T[K];
} &
  {
    [K in KeysWithOptionalValues<T>]?: Exclude<T[K], undefined>;
  };

/**
 * Validate that `value` is an object with string keys. The value at each key is tested against its own validator.
 *
 * Use this to create a predicate for a type like `{ a: string[] }`. For example:
 * ```ts
 * import * as schema from './schema';
 * const myValidator: (value: unknown) => value is { a: string[] } = schema.object({
 *   a: schema.array(schema.string()),
 * });
 * ```
 */
export function object<
  ValidatorDictionary extends {
    [key: string]: Validator<unknown>;
  }
>(
  validatorDictionary: ValidatorDictionary
): Validator<
  OptionalKeyWhenValueAcceptsUndefined<
    {
      [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]>;
    }
  >
> {
  valiator.explanation = function (value: unknown): Explanation {
    // This only validates non-null objects
    if (typeof value !== 'object' || value === null) {
      return [
        false,
        (newlinePrefix: string = '') =>
          `Expected a non-null object, but got ${JSON.stringify(value, null, newlinePrefix)}`,
      ];
    }

    // Rebind value as the result type so that we can interrogate it
    const trusted = value as { [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]> };

    const explanations: Array<[key: string, explanation: (newlinePrefix?: string) => string]> = [];

    // Get each validator in the validator dictionary and use it to validate the corresponding value
    for (const key of Object.keys(validatorDictionary)) {
      const result: Explanation = validatorDictionary[key].explanation(trusted[key]);
      if (!result[0]) {
        explanations.push([key, result[1]]);
      }
    }
    if (explanations.length) {
      return [
        false,
        (newlinePrefix: string = '') =>
          `Values in object were invalid. Explanations: ${explanations.map(
            ([key, explanation]) => `
${newlinePrefix + '\t'}* Value at key ${JSON.stringify(
              key,
              null,
              newlinePrefix + '\t'
            )}: ${explanation(newlinePrefix + '\t\t')} `
          )}`,
      ];
    } else {
      return [true, undefined];
    }
  };
  return valiator;
  function valiator(
    value: unknown
  ): value is /** If a key can point to `undefined`, then instead make the key optional and exclude `undefined` from the value type. */ OptionalKeyWhenValueAcceptsUndefined<
    {
      [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]>;
    }
  > {
    // This only validates non-null objects
    if (typeof value !== 'object' || value === null) {
      return false;
    }

    // Rebind value as the result type so that we can interrogate it
    const trusted = value as { [K in keyof ValidatorDictionary]: TypeOf<ValidatorDictionary[K]> };

    // Get each validator in the validator dictionary and use it to validate the corresponding value
    for (const key of Object.keys(validatorDictionary)) {
      const validator = validatorDictionary[key];
      if (!validator(trusted[key])) {
        return false;
      }
    }
    return true;
  }
}

/**
 * Validate that `value` is strictly equal to `acceptedValue`.
 * Use this for a literal type, for example:
 * ```
 * import * as schema from './schema';
 * const isAscOrDesc: (value: unknown) => value is 'asc' | 'desc' = schema.oneOf([
 *   schema.literal('asc' as const),
 *   schema.literal('desc' as const),
 * ]);
 * ```
 */
export function literal<T>(acceptedValue: T): Validator<T> {
  function validator(value: unknown): value is T {
    return validator.explanation(value)[0];
  }

  validator.explanation = function (value: unknown): Explanation {
    return value === acceptedValue
      ? [true, undefined]
      : [
          false,
          (newlinePrefix: string = '') =>
            `${JSON.stringify(value, null, newlinePrefix)} was expected to be ${JSON.stringify(
              acceptedValue,
              null,
              newlinePrefix
            )}`,
        ];
  };

  return validator;
}

/**
 * Validate that `value` is a string.
 * NB: this is used as `string` externally via named export.
 * Usage:
 * ```
 * import * as schema from './schema';
 * const isString: (value: unknown) => value is string = schema.string();
 * ```
 */
function anyString(): Validator<string> {
  validator.explanation = function (value: unknown): Explanation {
    if (typeof value !== 'string') {
      return [
        false,
        (newlinePrefix: string = '') =>
          `Expected a string, but got a ${typeof value}: ${JSON.stringify(
            value,
            null,
            newlinePrefix
          )}.`,
      ];
    } else {
      return [true, undefined];
    }
  };
  return validator;
  function validator(value: unknown): value is string {
    return validator.explanation(value)[0];
  }
}

/**
 * Validate that `value` is a number.
 * NB: this just checks if `typeof value === 'number'`. It will return `true` for `NaN`.
 * NB: this is used as `number` externally via named export.
 * Usage:
 * ```
 * import * as schema from './schema';
 * const isNumber: (value: unknown) => value is number = schema.number();
 * ```
 */
function anyNumber(): Validator<number> {
  validator.explanation = function (value: unknown): Explanation {
    return typeof value === 'number'
      ? [true, undefined]
      : [
          false,
          (newlinePrefix: string = '') =>
            `Expected a number, got a ${typeof value}: ${JSON.stringify(
              value,
              null,
              newlinePrefix
            )}`,
        ];
  };
  return validator;
  function validator(value: unknown): value is number {
    return validator.explanation(value)[0];
  }
}

/**
 * Export `anyString` as `string`. We can't define a function named `string`.
 * Export `anyNumber` as `number`. We can't define a function named `number`.
 */
export { anyString as string, anyNumber as number };
