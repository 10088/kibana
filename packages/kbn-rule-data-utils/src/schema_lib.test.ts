/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import * as schema from './schema_lib';
import { Validator } from './schema_lib';

interface SortDefinition {
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
}
describe(`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
}`, () => {
  let validator: Validator<SortDefinition>;
  beforeEach(() => {
    validator = schema.object({
      page: schema.oneOf([schema.literal(undefined), schema.number()]),
      sort: schema.oneOf([
        schema.array(
          schema.object({
            field: schema.oneOf([
              schema.literal('name' as const),
              schema.literal('ip' as const),
              schema.literal('date' as const),
            ]),
            direction: schema.oneOf([
              schema.literal('asc' as const),
              schema.literal('desc' as const),
            ]),
          })
        ),
        schema.literal(undefined),
      ]),
    });
  });
  type TestCaseDescriptor = [value: unknown, valid: boolean];
  const testCases: TestCaseDescriptor[] = [
    [
      {
        page: 1,
        sort: [
          {
            field: 'name',
            direction: 'asc',
          },
          {
            field: 'date',
            direction: 'desc',
          },
        ],
      },
      true,
    ],
    [
      {
        // page has the wrong type
        page: '1',
        sort: [
          {
            field: 'name',
            direction: 'asc',
          },
          {
            field: 'date',
            direction: 'desc',
          },
        ],
      },
      false,
    ],

    [
      {
        sort: [
          {
            // missing direction
            field: 'name',
          },
        ],
      },
      false,
    ],
    [
      {
        sort: [
          {
            field: 'name',
            // invalid direction
            direction: 'invalid',
          },
        ],
      },
      false,
    ],
    [
      {
        sort: [
          {
            // missing field
            direction: 'desc',
          },
        ],
      },
      false,
    ],
    [
      {
        sort: [
          {
            // invalid field
            field: 'invalid',
            direction: 'desc',
          },
        ],
      },
      false,
    ],
    // nothing in the array
    [{ sort: [] }, true],
    // page only
    [{ page: 1 }, true],
    // empty object (valid because all keys are optional.)
    [{}, true],
    // entirely invalid types
    [null, false],
    [true, false],
    ['', false],
  ];
  describe.each(testCases)('when the value to be validated is `%j`', (value, expected) => {
    it(`${expected ? 'should' : 'should NOT'} be valid`, () => {
      expect(validator(value)).toBe(expected);
    });
    if (expected) {
      it(`should not have any explanation`, () => {
        expect(validator.explanation(value)).toEqual([true, undefined]);
      });
    } else {
      it(`should have the expected explanation`, () => {
        const actual = validator.explanation(value);
        expect(actual[0]).toBe(false);
        expect(actual[1]!()).toMatchSnapshot();
      });
    }
  });
});
