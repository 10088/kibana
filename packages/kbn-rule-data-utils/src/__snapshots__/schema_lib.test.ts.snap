// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`""\` should have the expected explanation 1`] = `"Expected a non-null object, but got \\"\\""`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`{"page":"1","sort":[{"field":"name","direction":"asc"},{"field":"date","direction":"desc"}]}\` should have the expected explanation 1`] = `
"Values in object were invalid. Explanations: 
* Value at key \\"page\\": \\"1\\" was expected to match one of several conditions, but it did not. Here are the explanations: 
* \\"1\\" was expected to be undefined OR,
* Expected a number, got a string: \\"1\\" "
`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`{"sort":[{"direction":"desc"}]}\` should have the expected explanation 1`] = `
"Values in object were invalid. Explanations: 
* Value at key \\"sort\\": [{\\"direction\\":\\"desc\\"}] was expected to match one of several conditions, but it did not. Here are the explanations: 
* Values in array were invalid. The explanations: 
* Value at index 0: Values in object were invalid. Explanations: 
* Value at key \\"field\\": undefined was expected to match one of several conditions, but it did not. Here are the explanations: 
* undefined was expected to be \\"name\\" OR,
* undefined was expected to be \\"ip\\" OR,
* undefined was expected to be \\"date\\"  OR,
* [{\\"direction\\":\\"desc\\"}] was expected to be undefined "
`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`{"sort":[{"field":"invalid","direction":"desc"}]}\` should have the expected explanation 1`] = `
"Values in object were invalid. Explanations: 
* Value at key \\"sort\\": [{\\"field\\":\\"invalid\\",\\"direction\\":\\"desc\\"}] was expected to match one of several conditions, but it did not. Here are the explanations: 
* Values in array were invalid. The explanations: 
* Value at index 0: Values in object were invalid. Explanations: 
* Value at key \\"field\\": \\"invalid\\" was expected to match one of several conditions, but it did not. Here are the explanations: 
* \\"invalid\\" was expected to be \\"name\\" OR,
* \\"invalid\\" was expected to be \\"ip\\" OR,
* \\"invalid\\" was expected to be \\"date\\"  OR,
* [{\\"field\\":\\"invalid\\",\\"direction\\":\\"desc\\"}] was expected to be undefined "
`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`{"sort":[{"field":"name","direction":"invalid"}]}\` should have the expected explanation 1`] = `
"Values in object were invalid. Explanations: 
* Value at key \\"sort\\": [{\\"field\\":\\"name\\",\\"direction\\":\\"invalid\\"}] was expected to match one of several conditions, but it did not. Here are the explanations: 
* Values in array were invalid. The explanations: 
* Value at index 0: Values in object were invalid. Explanations: 
* Value at key \\"direction\\": \\"invalid\\" was expected to match one of several conditions, but it did not. Here are the explanations: 
* \\"invalid\\" was expected to be \\"asc\\" OR,
* \\"invalid\\" was expected to be \\"desc\\"  OR,
* [{\\"field\\":\\"name\\",\\"direction\\":\\"invalid\\"}] was expected to be undefined "
`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`{"sort":[{"field":"name"}]}\` should have the expected explanation 1`] = `
"Values in object were invalid. Explanations: 
* Value at key \\"sort\\": [{\\"field\\":\\"name\\"}] was expected to match one of several conditions, but it did not. Here are the explanations: 
* Values in array were invalid. The explanations: 
* Value at index 0: Values in object were invalid. Explanations: 
* Value at key \\"direction\\": undefined was expected to match one of several conditions, but it did not. Here are the explanations: 
* undefined was expected to be \\"asc\\" OR,
* undefined was expected to be \\"desc\\"  OR,
* [{\\"field\\":\\"name\\"}] was expected to be undefined "
`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`null\` should have the expected explanation 1`] = `"Expected a non-null object, but got null"`;

exports[`a validator made using the 'schema' module which validates that a value has the type:
{
  page?: number;
  sort?: Array<{ field: 'name' | 'ip' | 'date'; direction: 'asc' | 'desc' }>;
} when the value to be validated is \`true\` should have the expected explanation 1`] = `"Expected a non-null object, but got true"`;
